### PR TITLE
add vtadmin docker image

### PR DESCRIPTION
**Pull request by L3o-pold**
## Description

Add vtadmin docker image. Add the possibility to serve the vtadmin frontend inside docker image.

For now it's using node serving binary to serve the content but it could be change to have a single-component vtadmin.

Frontend have been modified in order to be able to change env var at the run time (not only build time). 

Usage in k8s looks like this:

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: vtadmin-web-config
data:
  config.js: |
    window.env = {
    'REACT_APP_VTADMIN_API_ADDRESS': "https://my-vtadmin-endpoint.local",
    'REACT_APP_FETCH_CREDENTIALS': "omit",
    'REACT_APP_ENABLE_EXPERIMENTAL_TABLET_DEBUG_VARS': true,
    'REACT_APP_BUGSNAG_API_KEY': "",
    'REACT_APP_DOCUMENT_TITLE': "",
    'REACT_APP_READONLY_MODE': false,
  };
```

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: vtadmin-api-config
data:
  discovery.json: |
    {
        "vtctlds": [
            {
                "host": {
                    "fqdn": "vtctld:15999",
                    "hostname": "vtctld:15999"
                }
            }
        ],
        "vtgates": [
            {
                "host": {
                    "hostname": "vtgate:15999"
                }
            }
        ]
    }
  rbac.yaml: |
    rules:
      - resource: "*"
        actions:
          - "get"
          - "create"
          - "delete"
          - "put"
          - "ping"
        subjects: ["*"]
        clusters: ["*"]
      - resource: "Shard"
        actions:
          - "emergency_reparent_shard"
          - "planned_reparent_shard"
        subjects: ["*"]
        clusters:
          - "cell1"

```

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: vtadmin
  labels:
    app.kubernetes.io/name: vtadmin
spec:
  replicas: 1
  selector:
    matchLabels:
      app.kubernetes.io/name: vtadmin
  template:
    metadata:
      labels:
        app.kubernetes.io/name: vtadmin
    spec:
      terminationGracePeriodSeconds: 30
      containers:
        - name: vtadmin
          image: vitess/vtadmin:v13.0.0-bullseye
          imagePullPolicy: IfNotPresent
          command:
            - /vt/bin/vtadmin
          args:
            - --addr=:14200
            - --http-origin=*
            - --cluster=id=cell1,name=cell1,discovery=staticfile,discovery-staticfile-path=/var/vtadmin/discovery.json,tablet-fqdn-tmpl={{ .Tablet.Hostname }}
            - --rbac
            - --rbac-config=/var/vtadmin/rbac.yaml
            - --tracer=opentracing-jaeger
            - --grpc-tracing
            - --http-tracing
            - --logtostderr
            - --alsologtostderr
          env:
            - name: GOMAXPROCS
              value: "4"
          ports:
            - containerPort: 14200
              name: tcp-api
              protocol: TCP
          resources:
            limits:
              cpu: 200m
              memory: 512Mi
            requests:
              cpu: 100m
              memory: 256Mi
          securityContext:
            runAsUser: 999
          volumeMounts:
            - mountPath: /var/vtadmin
              name: vtadmin-config
              readOnly: true
        - name: web
          image: vitess/vtadmin:v13.0.0-bullseye
          imagePullPolicy: IfNotPresent
          ports:
            - containerPort: 14201
              name: tcp-web
              protocol: TCP
          securityContext:
            runAsUser: 999
          volumeMounts:
            - mountPath: /vt/web/vtadmin/build/config
              name: vtadmin-web-config
              readOnly: true
          resources:
            limits:
              cpu: 200m
              memory: 512Mi
            requests:
              cpu: 100m
              memory: 256Mi
      volumes:
        - name: vtadmin-api-config
          configMap:
            name: vtadmin-api-config
        - name: vtadmin-web-config
          configMap:
            name: vtadmin-web-config
```

## Checklist

-   [x] "Backport me!" label has been added if this change should be backported
-   [x] Tests were added or are not required
-   [x] Documentation was added or is not required

### CODE REVIEW

File: docker/k8s/vtadmin/Dockerfile
Comment (derekperkins) [inline docker/k8s/vtadmin/Dockerfile:1]: 2022

File: docker/base/Dockerfile
Comment (ajm188) [inline docker/base/Dockerfile:59]: nit: newline at EOF please!

**Comment by L3o-pold** [comment]
yeah I saw where it was failing but no idea on how to fix it yet.

**Comment by notfelineit** [comment]
Hey @L3o-pold, have you tried making this a function instead of a var like so:
```
export const env = () => NodeJS.ProcessEnv = { ...window.env, ...process.env };
```
I'll take a deeper look at your PR as well, that's just a javascript-y hunch 😃

**Comment by L3o-pold** [comment]
image

My PR runs but I didn't try with the `REACT_APP_ENABLE_EXPERIMENTAL_TABLET_DEBUG_VARS` env var set to true.

**Comment by ajm188** [comment]
the problem is that `window.env` is not of type `NodeJS.ProcessEnv`, do we _need_ that variable, or can we type coerce it somehow so everything plays nice?

in any case, once we get that working, you're going to want to update the before/after mock infra (see http.test.ts for an example, which is explicitly mocking `process.env`, so you'll want to sub in your util function there instead)

**Comment by L3o-pold** [comment]
The goal of using `window.env` is to be able to edit environment variable as `REACT_APP_VTADMIN_API_ADDRESS` on the runtime. 

With this you can build a vtadmin dist version and allow people to edit those var when running the frontend. 

For example building a Vtadmin docker image and let users specify their own `REACT_APP_VTADMIN_API_ADDRESS` when running it.

**Comment by L3o-pold** [comment]
@ajm188 did a trick for fixing the CI. IMO cookie is ok only if we plan of adding an UI to edit those settings live. 
For now (with this PR), users need to play with `sed` in order to edit this settings during the runtime, but if we plan to replace `serve` dependency with go it could also be done directly in the binary I think.

**Comment by ajm188** [comment]
@notfelineit this[1] totally _would_ have worked, but there are tests for the error handler that mutate the `process.env` to make some assertions

[1] "this" meaning your suggestion here

> Are we able to access the variables we need simply by exporting an env var?
>
> `const env = {...window.env, ...process.env}`

**Comment by notfelineit** [comment]
@ajm188 could you link to the tests mutating process.env? That sounds like funny test behavior 🤔

**Comment by ajm188** [comment]
yep! https://github.com/vitessio/vitess/blob/main/web/vtadmin/src/api/http.test.ts#L58-L90 and also https://github.com/vitessio/vitess/blob/main/web/vtadmin/src/errors/errorHandler.test.ts#L21-L45

**Comment by notfelineit** [comment]
@ajm188 so I guess the issue with having a static object like `const env = {...window.env, ...process.env}` for the tests is that `env` might be set before the test process.env is applied? I don't see any tests that explicity make assertions against the process.env object, only the before/afters that set them.

I think a function would handle the above case ^ (recalculating `env` each time so it's always fresh). For handling type safety, would this work?:
```
const env = () => { ...window.env, ...process.env } as NodeJs.ProcessEnv
```

**Comment by ajm188** [comment]
that's ... exactly what i did. did you look at https://github.com/vitessio/vitess/pull/10543/commits/d8ca36a013978c5a34bb273d10f30ef8b0be7da3 ?

**Comment by notfelineit** [comment]
@ajm188 No I missed it! If that works for the tests, that sounds great

**Review by GuptaManan100 — COMMENTED** [review]
The changes to building the vtadmin web binaries are made in the `base` image which k8s is using so it works out there. But the `lite` images don't use `base`. They use `bootstrap` so this change may be needed in all of those images too.

File: docker/base/Dockerfile
Comment (deepthi) [inline docker/base/Dockerfile:34]: If we are adding this to the bootstrap image, then we don't need to add it here.
Instead, we need to first build and push bootstrap images with a new version number (10), and increment the version number in all the Dockerfiles that use that as the base.

**Review by deepthi — COMMENTED** [review]
Adding node to the `lite` image in order to run `vtadmin` is acceptable. We just need to measure the image size before and after and document the increase.

File: docker/k8s/vtadmin/Dockerfile
Comment (derekperkins) [inline docker/k8s/vtadmin/Dockerfile:48]: I don't understand why that needs to be writeable. Changes aren't persisted to the file, they are generally stored in the topo

Comment (derekperkins) [reply docker/k8s/vtadmin/Dockerfile:48]: Hmm... Shouldn't making it writable by the vitess user be sufficient like the other folders? IIUC, the entrypoint should be executed as that user

Comment (L3o-pold) [reply docker/k8s/vtadmin/Dockerfile:48]: no because we want to allow user to run the image with any user (like in the vitess operator with an non existing one)

File: docker/k8s/vtadmin/Dockerfile
Comment (derekperkins) [inline docker/k8s/vtadmin/Dockerfile:54]: What's the reason for having both `ENTRYPOINT` and `CMD`?

Comment (L3o-pold) [reply docker/k8s/vtadmin/Dockerfile:54]: It's easier to overwrite the command in a k8s deployment for example.

File: docker/bootstrap/Dockerfile.common
Comment (dbussink) [inline docker/bootstrap/Dockerfile.common:1]: This is using `bullseye`, but then we're using `buster` for the Go image. I don't think we should mix different bistro versions to avoid risk of mismatches in binary symbols etc.

File: docker/base/Dockerfile
Comment (deepthi) [reply docker/base/Dockerfile:34]: based on offline discussion, I now see that we are actually _building_ vtadmin in this step. So it doesn't belong in the bootstrap image at all.
bootstrap image should contain static dependencies (like node and npm). Builds need to happen in base and lite.
I was confused because the changes were also made to the bootstrap image.
One way forward is to revert changes to bootstrap image and make them in all the lite images. Then there is no need for a new bootstrap version either.
I'll leave it up to you, @ajm188 and @GuptaManan100 to decide whether node/npm need to be installed in the bootstrap image at all.

Comment (GuptaManan100) [inline docker/base/Dockerfile:34]: We should do the build of the vtadmin binaries in the k8s vtadmin docker file itself. We only use this build for that image, so it doesn't make sense to add it to all the other images.

File: docker/bootstrap/Dockerfile.common
Comment (GuptaManan100) [inline docker/bootstrap/Dockerfile.common:32]: This line should be removed. node is longer defined here. It will fail

**Review by GuptaManan100 — COMMENTED** [review]
Me and @deepthi had a discussion and I think for now it makes sense to only have the node installation, build, etc in the vtadmin image only. Let us not do this for the `lite` images until we have a clean way of doing this without any security concerns. 
This will allow us time after the release to evaluate our options and make changes without being rushed. Meanwhile we can introduce a new vtadmin image which can have node installed, do the build and the `sed` command changes, allowing us to deploy vtadmin in vtop so that people can start trying it out!

**Comment by ajm188** [comment]
@L3o-pold regarding the `sed` business and the need to `chmod 777` ...

i think instead we should provide a volume mount point, where the user can run a container with some sort of .env file (that has all the `REACT_VTADMIN_*` variables. Separately, we include a shell script in the image that, if that file is present, does the `sed`. This separates the image-build / container-init phases, doesn't require the arbitrary write perms, and solves the problem. Let me know what you think!

**Comment by L3o-pold** [comment]
@ajm188 not sure to understand. The sed requires the right on the `index.html` file. I don't see how providing a volume mount for `.env` file will solve this.

**Comment by L3o-pold** [comment]
I'll re-work to move from a `sed` / `index.html` thing to a `config.js` script loaded into the `index.html`. It will avoid any permission and weird entrypoint magic and will let the user mount his config file like (or if) he wants.

File: web/vtadmin/public/config.js
Comment (ajm188) [inline web/vtadmin/public/config.js:8]: nit: add a trailing newline please

File: docker/lite/Dockerfile.mysql57
Comment (deepthi) [inline docker/lite/Dockerfile.mysql57:54]: This reverts #9405 and probably breaks anyone currently relying on it.
Let's revert this one line for now.

**Comment by deepthi** [comment]
Since vtadmin is its own image, the example usage in the description needs to be changed to reflect that.
Have you verified that the example config actually works with a vtadmin image built with this PR?

**Review by deepthi — APPROVED** [review]
This is looking good to me. It looks like all feedback has been addressed
- no changes to bootstrap images
- no changes to lite images
- distro (bullseye) is not hard-coded, it is passed to the release script
- working example (confirmed with PR author via slack) in the description

If anyone finds problems with this after we merge it, we can fix it forward.

**Pull request merged**
